### ISSUE TITLE
Erroneous update-in-place on datadog_monitor_json with restriction policy

### ISSUE THREAD

**Issue by sscheible** [DataDog/terraform-provider-datadog#2515]
### Datadog Terraform Provider Version

v3.42.0

### Terraform Version

v1.9.0

### What resources or data sources are affected?

datadog_monitor_json

### Terraform Configuration Files

```hcl
resource "datadog_monitor_json" "sample_monitor" {
  monitor = <<-EOF
{
	"id": 18360519,
	"name": "Sample Monitor",
	"type": "log alert",
	"query": "logs(\"env:prod\").index(\"*\").rollup(\"count\").last(\"5m\") > 1",
	"message": "Sample Monitor is firing",
	"tags": [
		"team:datadog-project"
	],
	"options": {
		"thresholds": {
			"critical": 1
		},
		"enable_logs_sample": false,
		"notify_audit": false,
		"on_missing_data": "default",
		"include_tags": false,
		"new_host_delay": 300,
		"groupby_simple_monitor": false,
		"silenced": {}
	},
	"priority": null,
	"restriction_policy": {
		"bindings": [
			{
				"relation": "editor",
				"principals": [
					"team:67885f39-07e6-43bb-83a1-b12b656c41f5"
				]
			},
			{
				"relation": "viewer",
				"principals": [
					"org:4e66b06e-40d3-11ee-b8d2-da7ad0900005"
				]
			}
		]
	}
}
EOF
}
```

### Relevant debug or panic output

  # datadog_monitor_json.sample_monitor will be updated in-place
  ~ resource "datadog_monitor_json" "sample_monitor" {
        id      = "18360519"
      ~ monitor = jsonencode(
          ~ {
                name               = "Sample Monitor"
              + restriction_policy = {
                  + bindings = [
                      + {
                          + principals = [
                              + "team:67885f39-07e6-43bb-83a1-b12b656c41f5",
                            ]
                          + relation   = "editor"
                        },
                      + {
                          + principals = [
                              + "org:4e66b06e-40d3-11ee-b8d2-da7ad0900005",
                            ]
                          + relation   = "viewer"
                        },
                    ]
                }
                tags               = [
                    "team:datadog-project",
                ]
                # (5 unchanged attributes hidden)
            }
        )
    }

### Expected Behavior

tf plan should show no changes

### Actual Behavior

tf plan shows that terraform would add the restriction policy (which is already in place, confirmed via UI)

### Steps to Reproduce

1. terraform apply # (to create the monitor with restriction policy)
2. terraform plan # (to see if the plan differs from what has just been created)

### Important Factoids

Related resource [datadog_monitor](https://registry.terraform.io/providers/DataDog/datadog/3.42.0/docs/resources/monitor) does not support restriction policies at all, might be a related omission in the implementation within datadog_monitor_json

**Comment by hwangm** [comment]
I think I am seeing a similar issue, terraform v1.9.3 with dd provider 3.42.0. Every time I do a run on a project that has datadog_monitor_json resources, those resources are always being updated with the same metadata that is already in the state. It is as if the state isn't actually updated with the metadata. 

Module containing the json resource
```
resource "datadog_monitor_json" "error_tracking_monitors" {
  for_each = var.deployments

  monitor = <<-EOF
{
  "name": "[${var.environment}] New error in ${each.value}",
  "type": "error-tracking alert",
  "query": "error-tracking-traces(\"service:${each.value} env:${var.environment} @issue.age:<=300000\").rollup(\"count\").by(\"@issue.id\").last(\"5m\") > ${var.min_error_count}",
  "message": "Issue ID: {{[@issue.id].name}} \n\n<https://app.datadoghq.com/apm/error-tracking?query=%40issue.id%3A{{[@issue.id].name}}|View errors in Datadog>\n\n@${var.opsgenie_recipient}",
  "tags": [
    "env:${var.environment}",
    "service:${var.service}",
    "terraform:true"
  ],
  "priority": ${var.priority},
  "options": {
    "thresholds": {
      "critical": ${var.min_error_count},
      "period": {
        "digit": 5,
        "name": "5 minute average",
        "no_data_timeframe": 10,
        "seconds": 300,
        "tense": "last",
        "text": "5 minutes",
        "unit": "minutes",
        "value": "last_5m",
        "timeString": "5m",
        "unitAbbreviation": "m"
      },
      "timeAggregator": "avg",
      "comparison": ">"
    },
    "enable_logs_sample": true,
    "notify_audit": false,
    "notify_no_data": false,
    "new_host_delay": 300,
    "include_tags": true,
    "restriction_query": null,
    "escalation_message": "${var.escalation_message}"
  }
}
EOF
}
```

The plan
![Screenshot 2024-08-13 at 3 09 26 PM](https://github.com/user-attachments/assets/5296bbda-fe67-42a8-b835-064de5cf5b18)

In my case the threshold metadata seems to be rewritten on every run even though we are not making constant changes

**Comment by sa-dbilling** [comment]
We're also seeing this issue on the latest provider version 3.50.0

**Comment by AlexVasiukov** [comment]
This issue is still actual.

**Comment by jcpoconnor** [comment]
Seeing same issue as well.

**Comment by sscheible** [comment]
If I'm reading https://github.com/DataDog/terraform-provider-datadog/pull/2767 right (which was merged just after v3.53.0), next release v3.54.0 should have a fix

**Comment by AlexVasiukov** [comment]
@sscheible, as I see, #2767 was not merged. It was closed. So, the issue is still actual.

**Comment by AlexVasiukov** [comment]
@ecdatadog, @sanjaythakkar-dd,
Are there any expectations regarding the fix for this bug?  
It seems like a simple fix is required.  
However, the bug was reported in August last year and is still relevant.

**Comment by ecdatadog** [comment]
I agree the issue is relevant (so the issue is still opened) but we are still looking on how to fix properly the issue.

**Comment by sscheible** [comment]
> I agree the issue is relevant (so the issue is still opened) but we are still looking on how to fix properly the issue.

@ecdatadog any news? It's been a while, and the issue keeps polluting terraform plans to unreadability. Still actual with Terraform
v1.11.4 and datadog provider v3.63.0

**Comment by fpighi** [comment]
The issue should be resolved starting with version [3.91.0](https://github.com/DataDog/terraform-provider-datadog/blob/master/CHANGELOG.md#3910-march-4-2026).
